### ISSUE TITLE
Compiling errors on OS X 10.11

### ISSUE THREAD

**Issue by zhovner** [4ZM/mfterm#30]
```
$ ./configure 
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for bison... bison -y
checking for bison... (cached) bison -y
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking for flex... flex
checking for flex... (cached) flex
checking lex output file root... lex.yy
checking lex library... -ll
checking whether yytext is a pointer... yes
checking whether make sets $(MAKE)... (cached) yes
checking for ranlib... ranlib
checking for nfc_open in -lnfc... yes
checking for rl_completion_matches in -lreadline... yes
checking for DES_set_key_unchecked in -lcrypto... yes
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking stddef.h usability... yes
checking stddef.h presence... yes
checking for stddef.h... yes
checking for stdint.h... (cached) yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking for strings.h... (cached) yes
checking for stdbool.h that conforms to C99... yes
checking for _Bool... yes
checking for size_t... yes
checking for uint8_t... yes
checking for ptrdiff_t... yes
checking for stdlib.h... (cached) yes
checking for GNU libc compatible malloc... (cached) yes
checking for memset... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strdup... yes
checking for strtol... yes
checking for strtoul... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands

user:mfterm a$ make
\
	 \
	/bin/sh ./ylwrap `test -f 'spec_parser.y' || echo './'`spec_parser.y y.tab.c libsp_a-spec_parser.c y.tab.h `echo libsp_a-spec_parser.c | sed -e s/cc$/hh/ -e s/cpp$/hpp/ -e s/cxx$/hxx/ -e s/c++$/h++/ -e s/c$/h/` y.output libsp_a-spec_parser.output -- bison -y --name-prefix=sp_ -d 
/Users/user/mfterm/spec_parser.y:63.54-61: syntax error, unexpected type
make: *** [libsp_a-spec_parser.c] Error 1
```

**Comment by 4ZM** [comment]
Sorry for slow response to this one. I don't have access to a mac to debug this on, so not sure how to fix this. Tried compiling with clang / llvm on linux and that works with the latest from master.

Can you try adding #include <string.h> to spec_tokenizer.l and see it that fixes the problem?

**Comment by zhovner** [comment]
Get the same error.

I can give you remote access to mac.

**Comment by 4ZM** [comment]
I have just now committed a fix to master that I hope will fix this issue. Please let me know if it works.

**Comment by zhovner** [comment]
```
warning: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Wunknown-warning-option]
1 warning generated.
mv -f .deps/libdp_a-libdp_a-dictionary_parser.Tpo .deps/libdp_a-libdp_a-dictionary_parser.Po
rm -f libdp.a
ar cru libdp.a libdp_a-libdp_a-dictionary_parser.o 
ranlib libdp.a
\
	 \
	/bin/sh ./ylwrap `test -f 'spec_tokenizer.l' || echo './'`spec_tokenizer.l lex.yy.c libsp_a-spec_tokenizer.c -- flex --prefix=sp_ -olex.yy.c 
gcc -DHAVE_CONFIG_H -I.    -g -std=c99 -Wall -Wno-unused-function -Wno-unused-but-set-variable -Wno-implicit-function-declaration -g -O2 -MT libsp_a-libsp_a-spec_tokenizer.o -MD -MP -MF .deps/libsp_a-libsp_a-spec_tokenizer.Tpo -c -o libsp_a-libsp_a-spec_tokenizer.o `test -f 'libsp_a-spec_tokenizer.c' || echo './'`libsp_a-spec_tokenizer.c
warning: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Wunknown-warning-option]
1 warning generated.
mv -f .deps/libsp_a-libsp_a-spec_tokenizer.Tpo .deps/libsp_a-libsp_a-spec_tokenizer.Po
gcc -DHAVE_CONFIG_H -I.    -g -std=c99 -Wall -Wno-unused-function -Wno-unused-but-set-variable -Wno-implicit-function-declaration -g -O2 -MT libsp_a-libsp_a-spec_parser.o -MD -MP -MF .deps/libsp_a-libsp_a-spec_parser.Tpo -c -o libsp_a-libsp_a-spec_parser.o `test -f 'libsp_a-spec_parser.c' || echo './'`libsp_a-spec_parser.c
warning: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Wunknown-warning-option]
1 warning generated.
mv -f .deps/libsp_a-libsp_a-spec_parser.Tpo .deps/libsp_a-libsp_a-spec_parser.Po
rm -f libsp.a
ar cru libsp.a libsp_a-libsp_a-spec_tokenizer.o libsp_a-libsp_a-spec_parser.o 
ranlib libsp.a
gcc -DHAVE_CONFIG_H -I.    -g -Wall -Wconversion -std=c99 -Werror -g -O2 -MT mfterm.o -MD -MP -MF .deps/mfterm.Tpo -c -o mfterm.o mfterm.c
mfterm.c:170:3: error: use of undeclared identifier 'rl_completion_suppress_append'
  rl_completion_suppress_append = 0;
  ^
mfterm.c:174:5: error: use of undeclared identifier 'rl_completion_suppress_append'
    rl_completion_suppress_append = 1; // no trailing space on paths
    ^
2 errors generated.
make[2]: *** [mfterm.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

```

**Comment by 4ZM** [comment]
Jay! the patch worked. Thanks for helping me fix that. Alas, show me the joy that lasts...

This is another issue related to the readline library in your build environment. It seems that it is not compatible with the readline library installed on Linux.

What version of readline is installed on your computer? Please paste result of `brew info readline`

Also, could you paste the content of the file `/usr/include/readline/readline.h` from your system to pastebin and post a link to the paste?

Finally, can you do a search in the files in `/usr/include/readline` directory for the strings `rl_completion_suppress_append` and `rl_completion_suppress_append` to see if they turn up anywhere?

**Comment by zhovner** [comment]
```bash
$ brew info readline
readline: stable 6.3.8 (bottled) [keg-only]
```

```bash
$ grep -ir rl_completion_suppress_append /usr/include/* 
/usr/include/python2.6/pyconfig.h:/* Define if you have rl_completion_suppress_append */
/usr/include/python2.6/pyconfig.h:/* #undef HAVE_RL_COMPLETION_SUPPRESS_APPEND */
/usr/include/python2.7/pyconfig.h:/* Define if you have rl_completion_suppress_append */
/usr/include/python2.7/pyconfig.h:/* #undef HAVE_RL_COMPLETION_SUPPRESS_APPEND */
```

And here is  `/usr/include/readline/readline.h`:  

```C
/*	$NetBSD: readline.h,v 1.33 2012/05/15 17:30:04 christos Exp $	*/

/*-
 * Copyright (c) 1997 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Jaromir Dolecek.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#ifndef _READLINE_H_
#define _READLINE_H_

#include <sys/types.h>
#include <stdio.h>

/* list of readline stuff supported by editline library's readline wrapper */

/* typedefs */
typedef int	  Function(const char *, int);
typedef void	  VFunction(void);
typedef void	  VCPFunction(char *);
typedef char	 *CPFunction(const char *, int);
typedef char	**CPPFunction(const char *, int, int);
typedef char     *rl_compentry_func_t(const char *, int);
typedef int	  rl_command_func_t(int, int);

/* only supports length */
typedef struct {
	int length;
} HISTORY_STATE;

typedef void *histdata_t;

typedef struct _hist_entry {
	const char	*line;
	histdata_t 	 data;
} HIST_ENTRY;

typedef struct _keymap_entry {
	char type;
#define ISFUNC	0
#define ISKMAP	1
#define ISMACR	2
	Function *function;
} KEYMAP_ENTRY;

#define KEYMAP_SIZE	256

typedef KEYMAP_ENTRY KEYMAP_ENTRY_ARRAY[KEYMAP_SIZE];
typedef KEYMAP_ENTRY *Keymap;

#define control_character_threshold	0x20
#define control_character_bit		0x40

#ifndef CTRL
#include <sys/ioctl.h>
#if !defined(__sun) && !defined(__hpux) && !defined(_AIX)
#include <sys/ttydefaults.h>
#endif
#ifndef CTRL
#define CTRL(c)		((c) & 037)
#endif
#endif
#ifndef UNCTRL
#define UNCTRL(c)	(((c) - 'a' + 'A')|control_character_bit)
#endif

#define RUBOUT		0x7f
#define ABORT_CHAR	CTRL('G')
#define RL_READLINE_VERSION 	0x0402
#define RL_PROMPT_START_IGNORE	'\1'
#define RL_PROMPT_END_IGNORE	'\2'

/* global variables used by readline enabled applications */
#ifdef __cplusplus
extern "C" {
#endif
extern const char	*rl_library_version;
extern int 		rl_readline_version; 
extern char		*rl_readline_name;
extern FILE		*rl_instream;
extern FILE		*rl_outstream;
extern char		*rl_line_buffer;
extern int		 rl_point, rl_end;
extern int		 history_base, history_length;
extern int		 max_input_history;
extern char		*rl_basic_word_break_characters;
extern char		*rl_completer_word_break_characters;
extern char		*rl_completer_quote_characters;
extern Function		*rl_completion_entry_function;
extern char		*(*rl_completion_word_break_hook)(void);
extern CPPFunction	*rl_attempted_completion_function;
extern int		 rl_attempted_completion_over;
extern int		rl_completion_type;
extern int		rl_completion_query_items;
extern char		*rl_special_prefixes;
extern int		rl_completion_append_character;
extern int		rl_inhibit_completion;
extern Function		*rl_pre_input_hook;
extern Function		*rl_startup_hook;
extern char		*rl_terminal_name;
extern int		rl_already_prompted;
extern char		*rl_prompt;
/*
 * The following is not implemented
 */
extern KEYMAP_ENTRY_ARRAY emacs_standard_keymap,
			emacs_meta_keymap,
			emacs_ctlx_keymap;
extern int		rl_filename_completion_desired;
extern int		rl_ignore_completion_duplicates;
extern int		(*rl_getc_function)(FILE *);
extern VFunction	*rl_redisplay_function;
extern VFunction	*rl_completion_display_matches_hook;
extern VFunction	*rl_prep_term_function;
extern VFunction	*rl_deprep_term_function;
extern int		readline_echoing_p;
extern int		_rl_print_completions_horizontally;

/* supported functions */
char		*readline(const char *);
int		 rl_initialize(void);

void		 using_history(void);
int		 add_history(const char *);
void		 clear_history(void);
void		 stifle_history(int);
int		 unstifle_history(void);
int		 history_is_stifled(void);
int		 where_history(void);
HIST_ENTRY	*current_history(void);
HIST_ENTRY	*history_get(int);
HIST_ENTRY	*remove_history(int);
HIST_ENTRY	*replace_history_entry(int, const char *, histdata_t);
int		 history_total_bytes(void);
int		 history_set_pos(int);
HIST_ENTRY	*previous_history(void);
HIST_ENTRY	*next_history(void);
int		 history_search(const char *, int);
int		 history_search_prefix(const char *, int);
int		 history_search_pos(const char *, int, int);
int		 read_history(const char *);
int		 write_history(const char *);
int		 history_truncate_file (const char *, int);
int		 history_expand(char *, char **);
char	       **history_tokenize(const char *);
const char	*get_history_event(const char *, int *, int);
char		*history_arg_extract(int, int, const char *);

char		*tilde_expand(char *);
char		*filename_completion_function(const char *, int);
char		*username_completion_function(const char *, int);
int		 rl_complete(int, int);
int		 rl_read_key(void);
char	       **completion_matches(const char *, CPFunction *);
void		 rl_display_match_list(char **, int, int);

int		 rl_insert(int, int);
int		 rl_insert_text(const char *);
void		 rl_reset_terminal(const char *);
int		 rl_bind_key(int, rl_command_func_t *);
int		 rl_newline(int, int);
void		 rl_callback_read_char(void);
void		 rl_callback_handler_install(const char *, VCPFunction *);
void		 rl_callback_handler_remove(void);
void		 rl_redisplay(void);
int		 rl_get_previous_history(int, int);
void		 rl_prep_terminal(int);
void		 rl_deprep_terminal(void);
int		 rl_read_init_file(const char *);
int		 rl_parse_and_bind(const char *);
int		 rl_variable_bind(const char *, const char *);
void		 rl_stuff_char(int);
int		 rl_add_defun(const char *, Function *, int);
HISTORY_STATE	*history_get_history_state(void);
void		 rl_get_screen_size(int *, int *);
void		 rl_set_screen_size(int, int);
char 		*rl_filename_completion_function (const char *, int);
int		 _rl_abort_internal(void);
int		 _rl_qsort_string_compare(char **, char **);
char 	       **rl_completion_matches(const char *, rl_compentry_func_t *);
void		 rl_forced_update_display(void);
int		 rl_set_prompt(const char *);
int		 rl_on_new_line(void);

/*
 * The following are not implemented
 */
int		 rl_kill_text(int, int);
Keymap		 rl_get_keymap(void);
void		 rl_set_keymap(Keymap);
Keymap		 rl_make_bare_keymap(void);
int		 rl_generic_bind(int, const char *, const char *, Keymap);
int		 rl_bind_key_in_map(int, rl_command_func_t *, Keymap);
void		 rl_cleanup_after_signal(void);
void		 rl_free_line_state(void);
#ifdef __cplusplus
}
#endif

#endif /* _READLINE_H_ */
```

**Comment by 4ZM** [comment]
Looks like the readline header is some kind of wrapper of editline (some kind of more limited library) on OSX.

I've committed a patch to try to degrade readline (completion) support partially on missing library functions on some platforms. Please try the latest from master and `make distclean`, then the normal `./autogen.sh` and `./configure `.

**Comment by zhovner** [comment]
Finally works. I really enjoy your tool. Thank you for your great work. 

I have one error in make: 

```
mac.c:20:10: fatal error: 'openssl/des.h' file not found
#include <openssl/des.h>
```

i fixed it by `ln -s /usr/local/Cellar/openssl/1.0.2g/include/openssl /usr/local/include/openssl`

This issue described here 
https://github.com/phusion/passenger/issues/1630

**Comment by 4ZM** [comment]
Great! Thanks for your help troubleshooting this issue.
